lua/starfall/libs_cl/bass.lua

```diff
@@ -1,20 +1,107 @@
 local checkluatype = SF.CheckLuaType
 local registerprivilege = SF.Permissions.registerPrivilege
+local math_sqrt = math.sqrt
 
 -- Register privileges
 registerprivilege("bass.loadFile", "Play local sound files with `bass`.", "Allows users to create sound channels by file path.", { client = {} })
 registerprivilege("bass.loadURL", "Play remote sound files with `bass`.", "Allows users to create sound channels by URL.", { client = {}, urlwhitelist = {} })
 registerprivilege("bass.play2D", "Play sounds in global game context with `bass`.", "Allows users to create sound channels which play in global game context (without `3d` flag).", { client = { default = 1 } })
 
 local plyCount = SF.LimitObject("bass", "bass sounds", 20, "The number of sounds allowed to be playing via Starfall client at once")
+local soundDatas = {} -- { [IGModAudioChannel] = { ... }, ... } -- Contains extra data for each starfall sound.
+local simpleFadeSounds = {} -- { [1] = IGModAudioChannel, ... } -- A list of sounds that need to be manually controlled through 'simple fading'.
 
 SF.ResourceCounters.Bass = {icon = "icon16/sound_add.png", count = function(ply) return plyCount:get(ply).val end}
 
-local function deleteSound(ply, sound)
-	if sound:IsValid() then sound:Stop() end
+
+-- Calculates whether or not a 3D sound is out of earshot, and its fade multiplier.
+local function getSimpleFading(snd)
+	local pos = snd:GetPos()
+	local earPos = EyePos()
+	local fadeMin = snd:Get3DFadeDistance()
+	local fadeMax = soundDatas[snd].fadeMax
```

Comment:
soundData should be passed to this function instead of looking it up again. Also we could just store fadeMin and fadeMax both instead of using Get3DFadeDistance.